Add states DeltaRDriftInward, DeltaRNoDrift

markscheel opened the pull request:
These correspond to State 3 and State 4 in
SpEC.  These states can prevent the excision
boundary from getting too close to the horizon.

This involves:
 - Adding the two new states.
 - Adding new options DeltaRDriftInwardOptions
 - Passing in a new quantity into control_error.
 - Changing the three current states to allow transitions into the new states.
 - Adding new tests to Test_SizeControlStates that mirror the same tests in SpEC that test the logic of transitioning between states.
 - Updating the other tests to handle the new states.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
Add this option to your input file to have the DetalRDriftInward state disabled by default.
```yaml
ControlSystems:
  Size:
    ...
    DeltaRDriftInwardOptions: None
```
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel commented:
Failing test is Unit.PointwiseFunctions.AnalyticSolutions.Xcts.Schwarzschild (only for gcc-9 Release, passes all other cases). Something is really wrong with macOS x86; dies in spack.

wthrowe commented:
The macOS build is fixed in #6587, which I've just merged.

knelli2 commented on src/ControlSystem/ControlErrors/Size/DeltaRDriftOutward.cpp:
```diff
@@ -72,7 +72,8 @@ std::string DeltaRDriftOutward::update(
     ss << " Suggested timescale = " << info->suggested_time_scale;
   } else if (update_args.average_radial_distance.has_value() and
              update_args.average_radial_distance.value() <
-                 update_args.max_allowed_radial_distance.value()) {
+                 update_args.max_allowed_radial_distance.value_or(
+                     std::numeric_limits<double>::infinity())) {
```
This seems like a bug fix and not necessarily related to the other changes. Maybe move into it's own commit.

knelli2 commented on src/ControlSystem/ControlErrors/Size/DeltaRDriftInwardHelpers.hpp:
```diff
@@ -0,0 +1,42 @@
+namespace control_system::size::States {
+
+/// Value of target_char_speed when state DeltaRDriftInward is in effect.
+double target_speed_for_inward_drift(
```
I would just put all these helpers in the `DeltaRDriftInward.hpp` file. Everywhere you include the helpers you also include state 3 so seems like a logical place to put everything.

knelli2 commented on src/ControlSystem/ControlErrors/Size/DeltaRDriftInward.cpp:
```diff
@@ -0,0 +1,141 @@
+    // like we do in state DeltaR; this behavior agrees with SpEC.
+
+    // This factor prevents oscillating between states Initial and
+    // AhSpeed.  It needs to be slightly greater than unity, but the
```
Between Initial and AhSpeed? Or between DeltaR{Inward/Outward/NoDrif} and AhSpeed?

knelli2 commented on src/ControlSystem/ControlErrors/Size/DeltaRNoDrift.cpp:
```diff
@@ -0,0 +1,125 @@
+      // mode.
+
+      // This factor prevents oscillating between states Initial and
+      // AhSpeed.  It needs to be slightly greater than unity, but the
```
Same question about Initial

knelli2 commented on src/ControlSystem/ControlErrors/Size/Info.hpp:
```diff
@@ -79,6 +81,23 @@ struct CrossingTimeInfo {
+  /// if the minimum characteristic speed is increasing).
+  std::optional<double> t_drift_limit_char_speed;
+  /// t_delta_radius is the time (relative to the current time) when the
+  /// minimum distance between the horizon and the excision boundary is
```
`t_drift_limit_delta_radius`

knelli2 commented on src/ControlSystem/ControlErrors/Size/Info.hpp:
```diff
@@ -79,6 +81,23 @@ struct CrossingTimeInfo {
+  /// t_drift_limit_char_speed is the time (relative to the current time) when
+  /// the minimum characteristic speed is predicted to cross the
+  /// min_allowed_char_speed associated with state DeltaRDriftInward (or nullopt
+  /// if the minimum characteristic speed is increasing).
+  std::optional<double> t_drift_limit_char_speed;
```
It doesn't look like `t_drift_limit_char_speed` is used outside this class. Do you think we should still keep it exposed or just use it in the constructor to initialize `t_drift_limit` and discard?

markscheel replied:
I am removing it.

knelli2 commented on src/ControlSystem/ControlErrors/Size/DeltaRDriftInward.cpp:
```diff
@@ -0,0 +1,141 @@
+    info->state = std::make_unique<States::DeltaRNoDrift>();
+    info->suggested_time_scale = crossing_time_info.t_drift_limit;
+  } else if (crossing_time_info.t_delta_radius.has_value() and
+             info->damping_time >
+                 2.0 * spherepack_factor * update_args.horizon_00 * Y00) {
```
What's this factor of 2 for? This seems to be comparing the damping time to twice the average radius of the horizon

markscheel replied:
Adding a comment.

knelli2 commented on src/ControlSystem/ControlErrors/Size/State.hpp:
```diff
@@ -44,6 +47,26 @@ struct StateUpdateArgs {
+  /// Label::DeltaRDriftInward.
+  std::optional<double> min_allowed_radial_distance;
+  /// min_allowed_char_speed is the minimum char speed that will
+  /// trigger state Label::DeltaRDriftInward.  If both
```
I don't think you need `Label::` here

knelli2 commented on src/ControlSystem/ControlErrors/Size/DeltaR.cpp:
```diff
@@ -104,10 +118,27 @@ std::string DeltaR::update(const gsl::not_null<Info*> info,
              update_args.average_radial_distance.value() >
                  non_oscillation_drift_outward_factor *
-                     update_args.max_allowed_radial_distance.value()) {
+                     update_args.max_allowed_radial_distance.value_or(
+                         std::numeric_limits<double>::infinity())) {
```
This also seems like a bugfix that should go in a separate commit

knelli2 commented on src/ControlSystem/ControlErrors/Size/Initial.cpp:
```diff
@@ -59,25 +61,33 @@ std::string Initial::update(const gsl::not_null<Info*> info,
              update_args.average_radial_distance.value() >
                  non_oscillation_drift_outward_factor *
-                     update_args.max_allowed_radial_distance.value()) {
+                     update_args.max_allowed_radial_distance.value_or(
+                         std::numeric_limits<double>::infinity())) {
```
Also bugfix

knelli2 commented on src/ControlSystem/ControlErrors/Size/Initial.cpp:
```diff
@@ -59,25 +61,33 @@ std::string Initial::update(const gsl::not_null<Info*> info,
+                                     : "Comoving char speed positive")
+       << ". Switching to "
+       << (drift_inward ? "DeltaRDriftInward." : "DeltaR.")
+       << " with target char speed " << info->target_char_speed;
```
[nitpick] Print the char speed like the other times it's printed. i.e.

```cpp
   << ". Switching to "
   << (drift_inward ? "DeltaRDriftInward." : "DeltaR.") << "\n";
ss << " Target char speed = " << info->target_char_speed;
```

knelli2 commented on src/ControlSystem/ControlErrors/Size/Error.cpp:
```diff
@@ -186,10 +214,31 @@ ErrorDiagnostics control_error(
+          : std::nullopt;
 
   // Compute average radial distance for state DeltaRDriftOutward.
+  // NOTE: This choice corresponds to SpEC's "DeltaRPolicy=Absolute"
+  // and SpEC's "FunctionVsTimeMinDeltaRNoLam00=<NONE>".
+  // However, the default in SpEC is "DeltaRPolicy=Relative" and
+  // "FunctionVsTimeMinDeltaRNoLam00" being an actual ComputeItem.
```
So just for clarification, this means we won't be doing what spec does by default. But it's possible to add that in?

markscheel replied:
Yes, that could be added.  It could even be an option.  We may need an extra FunctionOfTime though, so I should think about how exactly to do that.

knelli2 commented on src/ControlSystem/ControlErrors/Size.hpp:
```diff
@@ -235,9 +235,47 @@ struct Size : tt::ConformsTo<protocols::ControlError> {
+    struct InwardDriftVelocity {
+      using type = double;
+      static constexpr Options::String help{
+          "Constant drift velocity term, if triggered by "
+          "MinAllowedRadialDistance."};
```
What velocity is used if triggered by `MinAllowedCharSpeed`? Or is this used for both? If I'm reading things right, the code does a minimum between this and another quantity but I don't think that needs to be explained in this help text.

markscheel replied:
Clarified help text.

knelli2 commented on src/ControlSystem/ControlErrors/Size/Error.hpp:
```diff
@@ -72,6 +77,19 @@ struct ErrorDiagnostics {
+ *        DeltaRDriftInward state.  If std::nullopt, then
+ *        DeltaRDriftInward will not be used.
+ * \param horizon_00 The l=0,m=0 component of the spherepack decomposition
+ *        of the apparent horizon.
```
Please make a note that you pass `horizon_00` separately from the full strahlkorper below because this one is time-averaged.

knelli2 commented on src/ControlSystem/ControlErrors/Size/DeltaRDriftInwardHelpers.hpp:
```diff
@@ -0,0 +1,42 @@
+/// Returs true if we should transition from state DeltaR to state
+/// DeltaRDriftInward.
+bool should_transition_from_state_delta_r_to_inward_drift(
+    const std::optional<double>& crossing_time_state_3,
+    double damping_time, const StateUpdateArgs& update_args);
+
+/// Returns true if we should transition from state DeltaRDriftInward
+/// to state DeltaRNoDrift.
+bool should_transition_from_state_inward_drift_to_delta_r_no_drift(
+    const std::optional<double>& crossing_time_state_3,
```
I think `crossing_time_state_3` should be either just `crossing_time` or `drift_crossing_time` (similar to the variable name that's usually passed in) so we avoid have `state_3` in any variable names.

markscheel replied:
Ok I am calling it crossing_time_drift_limit because it should be the same as info.t_drift_limit

knelli2 commented on tests/Unit/ControlSystem/ControlErrors/Test_SizeControlStates.cpp:
```diff
@@ -123,6 +153,97 @@ void do_test(const TestParams& test_params,
+  // should_transition_from_state_delta_r_to_inward_drift is true iff
+  // all of the following are true:
+  // A. should_activate_inward_drift is true
+  // B. crossing_time_state_3 >= damping time or crossing_time_state_3 == 0
```
Again with `state_3` in the name

knelli2 commented on tests/Unit/ControlSystem/ControlErrors/Test_SizeControlStates.cpp:
```diff
@@ -123,6 +153,97 @@ void do_test(const TestParams& test_params,
+  // all of the following are true:
+  // A. should_activate_inward_drift is true
+  // B. crossing_time_state_3 >= damping time or crossing_time_state_3 == 0
+  //    or inward_drift_velocity==0
```
You comments below only pertain to A. here (which combined 1-5 above). Is there anything we should know about B. in your comments below for each test?

markscheel replied:
Yes, added a comment that B. is satisfied on entry.  And simplified B. because if inward_drift_velocity doesn't have a value (which used to be part of B) then 1. is false.  Also changed the !=0 and ==0 to "has a value" and "does not have a value".  (the ==0 is how SpEC does it, and where I copied these comments from, but for SpECTRE these are std::optionals).

knelli2 reviewed (changes requested):
Looks good! Please add a fixup commit for the current changes and you can do the reordering while squashing.

As for organization I think we have two options, since I think you'll need a new commit for some bugfixes that were there before. Option A is

1. Commit for bugfixes/fixing existing clang-tidy issues
2. Commit for everything else

since everything is so closely coupled. Option B is

1. Commit for bugfixes/fixing existing clang-tidy issues
2. Commit that adds DeltaRDrifInward.?pp, DeltaRNoDrift.?pp, Info, State, and StateHistory. Basically everything needed for adding the state 3/4 files, but nothing needed for the logic of switching in the other states. No tests are added in this commit which should be stated in the message.
3. Commit that edits all the other states, adds the new options to Size control error, edits Error.?pp, Factory.hpp, register derived, input files, and adds the tests.

I only have a slight preference for option B, so if you have a strong aversion to doing the reordering that's ok and we can go with A.

markscheel commented:
Made changes in 2 fixup commits.
Did not split/reorder commits yet; will do when squashing.

markscheel commented:
Unit.Evolution.Systems.GhValenciaDivClean.TimeDerivativeTerms is the failed test on gcc-10 Debug.

markscheel committed:
Fix missing-value bugs in size control.

Under certain conditions, std::optional::value() could be
called on variables that had no value; now the code accounts
for the case in which there is no value.

markscheel committed:
Add states DeltaRDriftInward, DeltaRNoDrift

These correspond to State 3 and State 4 in
SpEC.  These states can prevent the excision
boundary from getting too close to the horizon.

NO TESTS are added in this commit; see the next commit.

The current states (Initial,DeltaRDriftOutward,AhSpeed,DeltaR) have
NOT been updated to transition to the new states in this commit.

This commit involves:
 - Adding the two new states
 - Updating Info to hold things that the new states use.
 - Updating State and StateHistory to know about the new states.

markscheel committed:
Integrate new Size states into control system.

The new states are DeltaRDriftInward and DeltaRNodrift.

This commit involves:
 - Adding new options DeltaRDriftInwardOptions.
 - Passing in a new quantity into the control error.
 - Changing the three existing states to allow transitions into
   the two new states.
 - Making Factory, RegisterDerived, and input files aware of the new
   states.
 - Adding new tests to Test_SizeControlStates that mirror the same
   tests in SpEC that test the logic of transitioning between states.
 - Updating the other tests to handle the new states.

knelli2 reviewed (commented):
Looks good! You can squash/rebase and reorder the commits. I still need to test on a singleBH and BBH run.

markscheel commented:
Ok, squashed and split commits!  (No, I'm not that fast. I already had done this on a local branch).

knelli2 reviewed (approved):
I've been trying to test this with a single BH, but haven't been able to actually get the control system to go into the DeltaRDrifInward state. We'll probably have to run more complicated binaries in order to test this.

For now, since this defaults to not using DeltaRDriftInward, I'll merge this and we can test as we go.
